codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/AwsJson.kt

```diff
@@ -183,159 +102,74 @@ class AwsJsonHttpBindingResolver(
-                         */
-                        parseErrorVariants(operationShape, errorSymbol)
-                    }
-                } else {
-                    write("return Err(#T::generic(generic))", errorSymbol)
-                }
-            }
-            val parser = generator.operationParser(operationShape)
-            Attribute.AllowUnusedMut.render(this)
-            rust("let mut builder = #T::default();", outputShape.builderSymbol(symbolProvider))
-            parser?.also {
-                rustTemplate(
-                    "builder = #{parse}(response.body().as_ref(), builder).map_err(#{error_symbol}::unhandled)?;",
-                    "parse" to it,
-                    "error_symbol" to errorSymbol
-                )
-            }
-            withBlock("Ok(builder.build()", ")") {
-                if (StructureGenerator.fallibleBuilder(outputShape, symbolProvider)) {
-                    rust(""".map_err(#T::unhandled)?""", errorSymbol)
-                }
-            }
-        }
-    }
+    override fun additionalHeaders(operationShape: OperationShape): List<Pair<String, String>> =
+        listOf("X-Amz-Target" to "${protocolConfig.serviceShape.id.name}.${operationShape.id.name}")
 
-    private fun fromResponseFun(
-        implBlockWriter: RustWriter,
-        operationShape: OperationShape,
-        block: RustWriter.() -> Unit
-    ) {
-        Attribute.Custom("allow(clippy::unnecessary_wraps)").render(implBlockWriter)
-        Attribute.AllowUnused.render(implBlockWriter)
-        implBlockWriter.rustBlock(
-            "fn parse_response(&self, response: & #T<#T>) -> std::result::Result<#T, #T>",
-            RuntimeType.Http("response::Response"),
-            RuntimeType.Bytes,
-            symbolProvider.toSymbol(operationShape.outputShape(model)),
-            operationShape.errorSymbol(symbolProvider)
-        ) {
-            block(this)
-        }
-    }
+    override fun structuredDataParser(operationShape: OperationShape): StructuredDataParserGenerator =
+        JsonParserGenerator(protocolConfig, httpBindingResolver)
+
+    override fun structuredDataSerializer(operationShape: OperationShape): StructuredDataSerializerGenerator =
+        AwsJsonSerializerGenerator(protocolConfig, httpBindingResolver)
 
-    private fun RustWriter.parseErrorVariants(
-        operationShape: OperationShape,
-        errorSymbol: RuntimeType
-    ) {
-        operationShape.errors.forEach { error ->
-            rustBlock("${error.name.dq()} => match #T(body)", RuntimeType.SerdeJson("from_value")) {
-                val variantName = symbolProvider.toSymbol(model.expectShape(error)).name
-                write(
-                    "Ok(body) => #1T { kind: #1TKind::$variantName(body), meta: generic },",
-                    errorSymbol
+    override fun parseGenericError(operationShape: OperationShape): RuntimeType {
+        return RuntimeType.forInlineFun("parse_generic_error", "json_deser") {
```

Comment:
nit: you can have this be a single Rust block which is typically a little easier to read